### PR TITLE
Move to use block based customisation for payloads

### PR DESCRIPTION
**Pull request by dwhenry**
This gives the use much greater flexibility without
try to cover every edge case, while still ensuring
the changes are validated against the schema.

This is a breaking change as it removes the existing
payload customisation methods.

https://trello.com/c/qx627h9B/396-fix-master-build

### CODE REVIEW

**Comment by tijmenb** [comment]
I'm starting to wonder if we're going too far with adding new methods to `customise_and_validate`. At some point it might be clearer to the user to generate a valid item, change it, and then validate it.

Maybe adding some sugar to do this makes it clearer?


```rb
random = GovukSchemas::RandomExample.for_schema("detailed_guide", schema_type: "frontend") do |content_item|
  content_item["details"] = "foo"
end
```

File: lib/govuk_schemas/random_example.rb [hunk 1]
```diff
@@ -40,93 +40,70 @@ def self.for_schema(schema_key_value)
     # Example:
     #
     #      random = GovukSchemas::RandomExample.for_schema("detailed_guide", schema_type: "frontend")
```
Comment (tijmenb) [inline lib/govuk_schemas/random_example.rb]: Should we make `for_schema` return the hash directly? We no longer need an object to talk to if it passes on the hash.

Comment (dwhenry) [reply lib/govuk_schemas/random_example.rb]: I can do. I had left that method so that existing implementation using the easy solution would not need to be changed.

Comment (dwhenry) [reply lib/govuk_schemas/random_example.rb]: Done

File: lib/govuk_schemas/version.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
 module GovukSchemas
   # @private
-  VERSION = "2.3.0".freeze
+  VERSION = "3.0.0".freeze
```
Comment (tijmenb) [inline lib/govuk_schemas/version.rb]: Can you do the version bump in a separate PR?

https://docs.publishing.service.gov.uk/manual/publishing-a-ruby-gem.html#versioning

Comment (dwhenry) [reply lib/govuk_schemas/version.rb]: Done.. and got the tests to pass again. :)

File: lib/govuk_schemas/random_example.rb [hunk 3]
```diff
@@ -29,9 +29,9 @@ def initialize(schema:)
...
     # @param schema_key_value [Hash]
     # @return [GovukSchemas::RandomExample]
-    def self.for_schema(schema_key_value)
+    def self.for_schema(schema_key_value, &block)
       schema = GovukSchemas::Schema.find(schema_key_value)
-      GovukSchemas::RandomExample.new(schema: schema)
+      GovukSchemas::RandomExample.new(schema: schema).payload(&block)
     end
 
     # Return a content item merged with a hash and with the excluded fields removed.
```
Comment (MatMoore) [inline lib/govuk_schemas/random_example.rb]: This description isn't accurate any more. It should probably describe the use case without a block first and then talk about passing a block.

File: lib/govuk_schemas/random_example.rb [hunk 4]
```diff
@@ -40,93 +40,70 @@ def self.for_schema(schema_key_value)
...
     #      # => {"base_path"=>"/e42dd28e", "title"=>"dolor est...", "publishing_app"=>"elit"...}
     #
-    # @param [Hash] hash The hash to merge the random content with
-    # @param [Array] array The array containing fields to exclude
+    #      random.payload do |payload|
+    #        payload['base_path'] = "Test base path"
+    #      end
+    #      # => {"base_path"=>"Test base path", "title"=>"dolor est...", "publishing_app"=>"elit"...}
+    #
+    # @param [Block] hash which can be modified before being validated
```
Comment (MatMoore) [inline lib/govuk_schemas/random_example.rb]: The wording is a bit strange here, it sounds a bit like the hash is the parameter.

Maybe it should say something like "an optional block that receives the generated content item and can modify it before it gets validated"

File: lib/govuk_schemas/random_example.rb (hunk 3 above)
Comment (dwhenry) [reply lib/govuk_schemas/random_example.rb]: updated

File: lib/govuk_schemas/random_example.rb (hunk 4 above)
Comment (dwhenry) [reply lib/govuk_schemas/random_example.rb]: updated

**Comment by dwhenry** [comment]
@tijmenb as discussed I have changed the code so the return value from the block is used in the validation step.
